### PR TITLE
PUD-852: Add MissingDocumentsRecord controller & persistence

### PR DESCRIPTION
**Pull request by philsegal**

### CODE REVIEW

File: src/main/resources/db/migration/V55__Add_missing_documents_record_tables.sql [hunk 1]
```diff
@@ -0,0 +1,15 @@
+CREATE TABLE missing_documents_record (
```
Comment (AndyNewton) [inline src/main/resources/db/migration/V55__Add_missing_documents_record_tables.sql:1]: File name needs to be at least "V56..." now.

File: src/main/resources/db/migration/V55__Add_missing_documents_record_tables.sql [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+          detail text NOT NULL,
+          version INTEGER NOT NULL,
+          created_by UUID NOT NULL,
+          created_date_time TIMESTAMPTZ NOT NULL
+);
+
+CREATE TABLE missing_document_category (
+    missing_document_record_id    UUID    NOT NULL REFERENCES missing_documents_record (id),
+    document_category   VARCHAR(32) NOT NULL,
+    UNIQUE (missing_document_record_id, document_category)
```
Comment (AndyNewton) [inline src/main/resources/db/migration/V55__Add_missing_documents_record_tables.sql:14]: I haven't spotted a test verifying this constraint.  For documents we have those in the IntegrationTest.  Tests named like "fun `cannot save and flush ... throws DataIntegrityViolationException`()".

**Review by AndyNewton — CHANGES_REQUESTED** [review]
Two minor QQs.  Let me know also when merge conflicts are resolved please.

File: src/main/resources/db/migration/V55__Add_missing_documents_record_tables.sql (hunk 2 above)
Comment (philsegal) [reply src/main/resources/db/migration/V55__Add_missing_documents_record_tables.sql:14]: we are saving a set for this. But for other restraints I will add.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/controller/MissingDocumentsRecordController.kt [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+data class MissingDocumentsRecordRequest(
+  val recallId: RecallId,
+  val categories: List<DocumentCategory>,
+  val detail: String,
+  val emailFileContent: String,
+  val emailFileName: String
+)
+
+data class MissingDocumentResponse(val documentId: DocumentId)
```
Comment (AndyNewton) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/controller/MissingDocumentsRecordController.kt:91]: Class unused?

File: src/main/resources/db/migration/V56__Add_missing_documents_record_tables.sql [hunk 4]
```diff
@@ -0,0 +1,16 @@
+CREATE TABLE missing_documents_record (
+          id UUID PRIMARY KEY,
+          recall_id UUID NOT NULL REFERENCES recall (id),
+          email_id UUID NOT NULL REFERENCES document (id),
+          detail text NOT NULL,
```
Comment (AndyNewton) [inline src/main/resources/db/migration/V56__Add_missing_documents_record_tables.sql:5]: 'TEXT' preferred to 'text' in general - later.

File: src/main/resources/db/migration/V56__Add_missing_documents_record_tables.sql [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+          version INTEGER NOT NULL,
+          created_by_user_id UUID NOT NULL,
+          created_date_time TIMESTAMPTZ NOT NULL,
+          UNIQUE (recall_id, version)
+);
+
+CREATE TABLE missing_document_category (
+    missing_document_record_id    UUID    NOT NULL REFERENCES missing_documents_record (id),
+    document_category   VARCHAR(32) NOT NULL,
```
Comment (AndyNewton) [inline src/main/resources/db/migration/V56__Add_missing_documents_record_tables.sql:14]: FYI the other store we have for the same info - I believe - is defined as `category  TEXT NOT NULL` so not as `VARCHAR`.
Of course there is a case for extracting those to a separate table too - but I don't think we need that for now.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/random/RandomUtilities.kt [hunk 6]
```diff
@@ -31,9 +31,15 @@ internal fun fullyPopulatedRecall(recallId: RecallId = ::RecallId.random()): Rec
...
-          // ensure document version is valid verus category
+          // ensure document version is valid versus category
           version = if (document.category.versioned) Random.nextInt() else null
         )
+      }.toSet(),
+      missingDocumentsRecords = recall.missingDocumentsRecords.map { mdr ->
+        mdr.copy(
+          recallId = recallId.value,
+          emailId = recall.documents.random().id
+        )
```
Comment (AndyNewton) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/random/RandomUtilities.kt:42]: Have you seen any instability in tests using `fullyPopulatedRecall`?   The comment `// ensure document version is valid versus category` and code for that is something I needed to add for that with versioned documents.

**Review by AndyNewton — DISMISSED** [review]
Maybe take a quick look at my latest comments - before or after merge, as you wish.

**Review by AndyNewton — APPROVED** [review]

**Pull request merged**